### PR TITLE
RFC: Keras SavedModel Update

### PR DESCRIPTION
**Pull request by k-w-w**
Comment Period will end 9/2/2020

# Keras SavedModel Update

| Status        | Proposed                                             |
:-------------- |:---------------------------------------------------- |
| **RFC #**     | [281](https://github.com/tensorflow/community/pull/281)|
| **Author(s)** | Kathy Wu (kathywu@google.com)                        |
| **Sponsor**   | Karmel Allison (karmel@google.com)                   |
| **Updated**   | 2020-08-20                                           |
| **Obsoletes** | Obsoletes some of the implementation details in [20190509-keras-saved-model.md](https://github.com/tensorflow/community/blob/master/rfcs/20190509-keras-saved-model.md)          |

## Objective

The goal of this proposal is to separate core SavedModel and Keras Model serialization. 
By doing so, Keras will only rely on external API when saving and loading models from the
SavedModel format. This allows us to define what the process of serializing and
deserializing high-level objects should be, in clear and unconvoluted terms. 

This document does not solve existing problems in the Keras model serialization, such as dealing with shared 
variables between layers.

This document introduces a change that may impact some users. Please see the [Breaking Changes](https://github.com/k-w-w/community/blob/patch-2/rfcs/20200809-keras-saved-model-update.md#breaking-changes) section for details.

### CODE REVIEW

File: rfcs/20200809-keras-saved-model-update.md [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+An alternative that was long considered was having the user register serializer and deserializer functions to each class that should be specially handled during SavedModel serialization and deserialization. However, this idea was put to rest due to the recursive nature of the TensorFlow object graph and the complicated issues [mentioned above](#why-separate-core-vs-high-level-serialization).
+
+### Breaking changes
+`tf.saved_model.save` will no longer save the metadata added to Keras models. With this, only the Keras `model.save` function will create SavedModels that are compatible with the Keras model loader `tf.keras.models.load_model`. When using the core loader (`tf.saved_model.load`), an object with specific endpoints is loaded. Please see the Keras SavedModel RFC for the list of all attributes that are stored: https://github.com/k-w-w/community/blob/master/rfcs/20190509-keras-saved-model.md#serialization-details
+
+| Saver  | Loader |  |
+| ---- | ---- | ---- | 
+| `tf.saved_model.save` | `tf.saved_model.load` | No change |
+| `tf.keras.models.save_model` | `tf.saved_model.load` | No change |
+| `tf.saved_model.save` | `tf.keras.models.load_model` | **Error** |
```
Comment (saberkun) [inline rfcs/20200809-keras-saved-model-update.md:134]: Does it mean tf.saved_model.save() will no longer trace each layer call() and enforce requirements?
Is there a support for multiple input arguments to model and layer?

File: rfcs/20200809-keras-saved-model-update.md [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+| `tf.keras.models.save_model` | `tf.keras.models.load_model` | No change |
+
+
+#### TF Hub compatibility
```
Comment (saberkun) [inline rfcs/20200809-keras-saved-model-update.md:138]: Besides TF Hub, can we discuss what are the changes to tf-serving? What are the benefits?
Is that true the savedmodel size will be smaller and we only trace the tf.function to export?
Will the speed of saving be faster?

Comment (k-w-w) [reply rfcs/20200809-keras-saved-model-update.md:138]: The SavedModel will still trace all tf.functions when saving a Keras layer (these are used by tf.Hub). There will be additional options to saving an inference-only SavedModel but that is not addressed here.

File: rfcs/20200809-keras-saved-model-update.md (hunk 1 above)
Comment (k-w-w) [reply rfcs/20200809-keras-saved-model-update.md:134]: tf.saved_model will still save each layer call by default -- only the metadata is removed. The bug involving multiple input arguments is being looked at but not in the scope of this doc

**Comment by bhack** [comment]
What is the status of this? I don't know if this is also will cover the issue about optimizer with slots. In this case I want to add a warning somewhere in the meantime https://github.com/tensorflow/tensorflow/pull/42846

File: rfcs/20200809-keras-saved-model-update.md [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+* Layer objects must be loaded first, since loading from the config may cause the creation of other nodes in the ObjectGraph.
+* Functional and Sequential models must be loaded after the layers, to reconnect the model layer network.
+* Any objects *not* created during the above steps should still be loaded. Meaning all objects created by a config must be tracked.
+* Dealing with shared objects that can be created by multiple constructors raises lots of complications.
+
+Thus, introducing a new deserializer API that addresses all these concerns is likely to be complicated and clunky. The core SavedModel API should focus on a single task -- that is saving/loading an object graph with low-level objects. Having a solid base API makes it easier to define serialization for high-level frameworks.
+
+### What should high-level serialization look like?
+High level libraries should utilize the core tf.saved_model.save and tf.saved_model.load functions to save and load the TensorFlow object graph and functions. Anything that is not saved or loaded by the core functions should be saved and loaded separately by the library.
```
Comment (adriangb) [inline rfcs/20200809-keras-saved-model-update.md:40]: > Anything that is not saved or loaded by the core functions should be saved and loaded separately by the library

Just to clarify, does "high level libraries" here refer to `tf.keras.Model.save` or "any 3rd party libraries that may want to save/load `tf.keras.Model` objects"?

**Comment by bhack** [comment]
What is the status of this RFC?

**Comment by ematejska** [comment]
This didn't seem to have gone through a design review meeting.  Following up.

**Comment by k-w-w** [comment]
Hi @bhack @ematejska , there was no design meeting for this change, but most of the changes (minus the removal of the metadata in the core SavedModel proto) has been submitted in TF 2.5. A warning is logged if you try to load a SavedModel that was saved prior to TF 2.5.

The metadata deprecation has not happened yet, so you can still save a SavedModel that is compatible with previous TensorFlow versions when calling `tf.saved_model.save`. You can expect by 2.6 the object metadata will be removed from the SavedModel.

**Comment by ematejska** [comment]
@k-w-w Should this RFC be closed out?

**Comment by ematejska** [comment]
@k-w-w This will be closed out soon if there is no response.

**Comment by k-w-w** [comment]
This can be closed

**Pull request closed without merging**